[FIX] stock_full_location_reservation: Don't unlink moves if a cancel is called

rousseldenis opened the pull request.

rousseldenis committed:
[FIX] stock_full_location_reservation: Don't unlink moves if a cancel is called

rousseldenis commented:
@mt-software-de @jbaudoux

mt-software-de commented on stock_full_location_reservation/models/stock_move.py:
```diff
@@ -13,6 +13,22 @@ class StockMove(models.Model):
 
+    def _action_cancel(self):
+        """
+        Override the method to do the unlink only as the super() do
+        the unreserve also.
```
Could we change the doc string to make it more clear? 
```suggestion
        Override the method to avoid the unlink of replenishment moves
        via _do_unsreserve which is getting called in the super
        Instead do the unlink after the super if needed
```

mt-software-de commented on stock_full_location_reservation/models/stock_move.py:
```diff
@@ -13,6 +13,22 @@ class StockMove(models.Model):
+        Override the method to do the unlink only as the super() do
+        the unreserve also.
+        """
+        # If we are already in a undo_full_location_reservation context,
+        # we return super() method.
+        if self.env.context.get("skip_undo_full_location_reservation"):
+            return super()._action_cancel()
+        new_self = self.with_context(skip_undo_full_location_reservation=True)
+        res = super(StockMove, new_self)._action_cancel()
+        full_location_moves = self._filter_full_location_reservation_moves()
+        if full_location_moves:
+            full_location_moves.unlink()
+        return res
```
```suggestion
        # If we are already in a undo_full_location_reservation context,
        # we return super() method.
        skip_undo = self.env.context.get("skip_undo_full_location_reservation")
        self = self.with_context(skip_undo_full_location_reservation=True)
        res = super()._action_cancel()
        if not skip_undo:
            full_location_moves = self._filter_full_location_reservation_moves()
            if full_location_moves:
                full_location_moves.unlink()
        return res
```

rousseldenis replied:
Ok, apart a remark I have already seen in your code. You cannot replace self. It's better to use a new variable.

mt-software-de replied:
Why? It is working for me. 
But this is not a blocker. :-) I'm fine with `new_self`

rousseldenis replied:
> Why? It is working for me.

Yes it works, but semantically it is not quite good. As self is not the same as the one at the function beginning. When reading code, you should be able to identify that 'self' is not what you expect to be.

mt-software-de replied:
Ok yes makes sense. :-)

mt-software-de reviewed (changes requested):
Minor remarks.

rousseldenis committed:
[IMP] stock_full_location_reservation: Improve docstring

rousseldenis committed:
[FIX/IMP] stock_full_location_reservation: Don't unlink moves at cancel

As _action_cancel() should not delete records, we should do the unlink()
asynchronuously. Do it as cron job instead

rousseldenis committed:
[FIX/IMP] stock_full_location_reservation: Do the unlinks as queue jobs

To get almost the first behavior (having the moves deleted at full unreservation),
do the unlinks as queue job.

rousseldenis commented:
@jbaudoux @mt-software-de I've update this as unlinking moves in the unreserve (as unreserve can call _action_cancel()) is a nonsense as records should be accessible later in the call stack.

To keep the current behavior, I've done the unlink() in a queue job in order to get the deletion asynchronuous.

rousseldenis committed:
[IMP] stock_full_location_reservation: Allow to not unreserve moves

Use the standard mechanism to update the reserved quantity instead of recreating
a new move (and move lines).

rousseldenis commented:
@mt-software-de

rousseldenis commented:
I've included this one (I have conflicts):

https://github.com/mt-software-de/wms/pull/7

rousseldenis committed:
[IMP] stock_full_location_reservation: Don't merge if products are different

rousseldenis committed:
[IMP] stock_full_location_reservation: Introduce a strict mode

As in some flows we need to limit the full reservation to the move
line characteristics (lot, package, owner), add a strict mode that
will use the Odoo core mechanism to gather the needed quant(ities).
